src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -164,20 +163,60 @@ public async Task<BundleResponse> Handle(BundleRequest bundleRequest, Cancellati
                     Type = BundleType.TransactionResponse,
                 };
 
-                return await ExecuteTransactionForAllRequests(responseBundle);
+                return await ExecuteTransactionForAllRequests(bundleResource, responseBundle, cancellationToken);
             }
 
             throw new MethodNotAllowedException(string.Format(Api.Resources.InvalidBundleType, _bundleType));
         }
 
-        private async Task<BundleResponse> ExecuteTransactionForAllRequests(Hl7.Fhir.Model.Bundle responseBundle)
+        private async Task ResolveVersionedReferencesAsync(Hl7.Fhir.Model.Bundle requestBundle, Hl7.Fhir.Model.Bundle responseBundle, Dictionary<string, (string resourceId, string resourceType)> referenceIdDictionary, CancellationToken cancellationToken)
+        {
+            // this will track the corresponding entry in request bundle.
+            int iterator = 0;
+
+            foreach (EntryComponent entry in responseBundle.Entry)
+            {
+                if (requestBundle.Entry[iterator].Request.Method == HTTPVerb.POST || requestBundle.Entry[iterator].Request.Method == HTTPVerb.PUT)
+                {
+                    IEnumerable<ResourceReference> references = entry.Resource.GetAllChildren<ResourceReference>();
+
+                    foreach (ResourceReference reference in references)
+                    {
+                        if (referenceIdDictionary.ContainsKey(reference.Reference) && IsVersionedUrl(reference.Reference))
+                        {
+                            // Extract the referencedResourceUrl from versionedReferenceUrl
+                            string referencedResourceUrl = reference.Reference.Split("/_history/")[0];
+                            string[] referencedResourceInformation = referencedResourceUrl.Split("/");
+                            string referencedResourceType = referencedResourceInformation[0];
+                            string referencedResourceId = referencedResourceInformation[1];
```

Comment:
Not sure, declaring extra variables referencedResourceType and referencedResourceId is better than using referencedResourceInformation[] directly. May be for readability we can put a comment. Thoughts?